mcserver/views.py

```diff
@@ -167,49 +188,75 @@ def get_queryset(self):
         if user.is_authenticated and user.id == 1:
             return Session.objects.all().order_by("-created_at")
         return Session.objects.filter(Q(user__id=user.id) | Q(public=True)).order_by("-created_at")
-       
+
     @action(detail=False)
     def api_health_check(self, request):
-        
         return Response({"status": "True"})
-    
+
     @action(
         detail=True,
-        methods=["get","post"],
+        methods=["get", "post"],
     )
     def calibration(self, request, pk):
-        session_path = "/data/{}".format(pk)
-        session = Session.objects.get(pk=pk)
-        trial = session.trial_set.filter(name="calibration").order_by("-created_at")[0]
+        try:
+            if pk == 'undefined':
+                raise ValueError(_("undefined_uuid"))
+
+            session = Session.objects.get(pk=pk)
+            trial = session.trial_set.filter(name="calibration").order_by("-created_at")[0]
 
-        trial.meta = {
-            "calibration": {
-                cam: val for cam, val in request.data.items()
+            trial.meta = {
+                "calibration": {
+                    cam: val for cam, val in request.data.items()
+                }
             }
-        }
-        trial.save()
+            trial.save()
+        except Http404:
+            if settings.DEBUG:
+                raise Exception(_("error") % {"error_message": str(traceback.format_exc())})
+            raise NotFound(_("session_uuid_not_found") % {"uuid": str(pk)})
+        except ValueError:
+            if settings.DEBUG:
+                raise APIException(_("error") % {"error_message": str(traceback.format_exc())})
+            raise NotFound(_("session_uuid_not_valid") % {"uuid": str(pk)})
+        except Exception:
+            if settings.DEBUG:
+                raise Exception(_("error") % {"error_message": str(traceback.format_exc())})
+            raise APIException(_("calibration_error"))
 
         return Response({
             "status": "ok",
             "data": request.data,
         })
 
-
     @action(detail=True, methods=['post'])
     def rename(self, request, pk):
         # Get session.
         session = get_object_or_404(Session.objects.all(), pk=pk)
 
         try:
+            if pk == 'undefined':
+                raise ValueError(_("undefined_uuid"))
             error_message = ""
 
             # Update session name and save.
             session.meta["sessionName"] = request.data['sessionNewName']
             session.save()
 
-        except Exception as e:
-            error_message = 'There was an error while renaming your session: ' + str(e)
-            print(error_message)
+            self.check_object_permissions(self.request, session)
```

Comment:
You are right, we should check for permissions before doing the action, I am going to change it everywhere it should be done.

Regarding the error, yes, it should return 401 Unauthorized, instead of executing the rest of the code. But if you check the except blocks, there is no specific case for it, so it will enter into the generic Exception case. I have added specific cases for this when we check for permissions.